components/script_plugins/unrooted_must_root.rs

```diff
@@ -139,97 +143,105 @@ impl<'a, 'tcx> LateLintPass<'a, 'tcx> for UnrootedPass {
+                        false
+                    } else if did.is_box() && in_new_function {
+                        // box in new() is okay
+                        false
+                    } else {
+                        true
+                    }
+                },
+                ty::TyParam(param_ty) => {
+                    let ty_param_def = cx.tcx.generics_of(self.def_id).type_param(&param_ty, cx.tcx);
+                    // will work for `foo::<Foo>(..)`
+                    // but won't get enough information in `bar<T>(){foo::<T>(..);}`?
+                    if cx.tcx.has_attr(ty_param_def.def_id, "must_root") {
+                        ret = true;
+                        false
+                    } else {
+                        true
+                    }
+                },
+                ty::TyRef(..) => false, // don't recurse down &ptrs
+                ty::TyRawPtr(..) => false, // don't recurse down *ptrs
+                ty::TyFnDef(..) | ty::TyFnPtr(_) => false,
+                _ => true
             }
-        }
-
-        visit::walk_expr(self, expr);
+        });
+        ret
     }
+}
 
-    fn visit_pat(&mut self, pat: &'tcx hir::Pat) {
-        let cx = self.cx;
-
-        // We want to detect pattern bindings that move a value onto the stack.
-        // When "default binding modes" https://github.com/rust-lang/rust/issues/42640
-        // are implemented, the `Unannotated` case could cause false-positives.
-        // These should be fixable by adding an explicit `ref`.
-        match pat.node {
-            hir::PatKind::Binding(hir::BindingAnnotation::Unannotated, _, _, _) |
-            hir::PatKind::Binding(hir::BindingAnnotation::Mutable, _, _, _) => {
-                let ty = cx.tables.pat_ty(pat);
-                if is_unrooted_ty(cx, ty, self.in_new_function) {
-                    cx.span_lint(UNROOTED_MUST_ROOT,
-                                pat.span,
-                                &format!("Expression of type {:?} must be rooted", ty))
-                }
+impl<'a, 'b, 'tcx> MirVisitor<'tcx> for UnrootedCx<'a, 'b, 'tcx> {
+    fn visit_local_decl(&mut self, _local: mir::Local, decl: &mir::LocalDecl<'tcx>) {
+        if decl.is_user_variable {
+            if self.is_unrooted_ty(decl.ty, self.in_new_function) {
+                self.late_cx.span_lint(UNROOTED_MUST_ROOT, decl.source_info.span, "Type of binding/expression must be rooted.")
             }
-            _ => {}
         }
-
-        visit::walk_pat(self, pat);
-    }
-
-    fn visit_ty(&mut self, _: &'tcx hir::Ty) {}
-
-    fn nested_visit_map<'this>(&'this mut self) -> hir::intravisit::NestedVisitorMap<'this, 'tcx> {
-        hir::intravisit::NestedVisitorMap::OnlyBodies(&self.cx.tcx.hir)
+        // how to check return value and arguments?
+        // unpack local and check the number? (0=ret place, 1..arg_count = args?) how to nicely pass arg_count/mir to this method?
```

Comment:
.

Reply:
@nikomatsakis I haven't removed the check from `check_fn` cause I don't know how to nicely move it here.

Reply:
I would put the `&Mir` into the context and then use `mir.local_kind()`